manager/controlapi/volume.go

```diff
@@ -0,0 +1,226 @@
+package controlapi
+
+import (
+	"context"
+
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/identity"
+	"github.com/docker/swarmkit/manager/state/store"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+func (s *Server) CreateVolume(ctx context.Context, request *api.CreateVolumeRequest) (*api.CreateVolumeResponse, error) {
+	if request.Spec == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "spec must not be nil")
+	}
+
+	// validate the volume spec
+	if request.Spec.Driver == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "driver must be specified")
+	}
+
+	if request.Spec.Annotations.Name == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "meta: name must be provided")
+	}
+
+	volume := &api.Volume{
+		ID:   identity.NewID(),
+		Spec: *request.Spec,
+	}
+	err := s.store.Update(func(tx store.Tx) error {
+		// verify that the specified driver exists
+		clusters, err := store.FindClusters(tx, store.ByName(store.DefaultClusterName))
+		if err == nil && len(clusters) == 1 {
```

Comment:
To clarify, is the significance of looking for only one cluster because there should only be one `DefaultClusterName`?

Reply:
Yes. For Very Stupid Reasons™, the system supports more than one cluster object. Only one is ever created, and it always uses `DefaultClusterName`. But the ID can be different on different clusters, so we have to `Find` and not `Get`